### ISSUE TITLE
Compatibility with :share()'d layers

### ISSUE THREAD

**Issue by achalddave** [gpleiss/nnlr#1]
nnlr is super useful, thanks for the package!

I was able to use it with a few networks, but I'm having trouble with networks where I'm sharing weights across layers. Specifically, if I have two layers with `n` parameters each that are shared, it seems `:getParameters()` returns a vector of size `n`, while `getOptimConfig` returns vectors of length `2*n`. This causes issues with `optim`, which expects the parameters and learning rates vector to be of the same length. 

Any help would be appreciated - thanks again!

~~Edit: On second thought, I'm not sure this is the issue I'm running into. The general issue is that the output of `getParameters` is smaller than that of `getOptimConfig` in size, and this only happened in a network with `share()`'d layers. But perhaps the origin of the issue is something else.~~
Update: The difference in size between `getParameters` and `getOptimConfig` is exactly the number of shared parameters, so I'm fairly certain the issue is with `nnlr` and `share()`'d layers.

**Comment by gpleiss** [comment]
Hmm... I'll investigate this! This might be a case I forgot to account for.
Hopefully we can get this fixed soon!

**Comment by achalddave** [comment]
Yeah unfortunately I can't come up with how to do this if `:share` was called before `nnlr` was `require`d. If `share` was called after `nnlr` is imported, presumably you could override `share` so that `share('weight')` would cause you to share the weight learning rates and weight weight decays.

Edit: At the cost of `clone`ing the model, perhaps something like https://github.com/torch/nn/issues/466#issuecomment-220137220 could be used directly instead? That is, when `getOptimConfig` is called, clone the model, set the weight to be the learning rates, then call `getParameters`. This can be repeated for weight decays. This result can be cached until any learning rates or weight decays are updated. If this sounds right, I can take a shot at implementing it.

**Comment by gpleiss** [comment]
Ugh sorry for the super long delay.
I think what you suggest should work! I'm hoping there's a slightly less hacky solution :)
I'll play around with it a bit myself, but I will happily accept a PR if your way works!
